Add various property validation in EmbedBuilder

moiph opened the pull request.

moiph committed:
Add various property validation in EmbedBuilder

moiph commented on src/Discord.Net.Rest/Entities/Messages/EmbedBuilder.cs:
```diff
@@ -8,14 +8,36 @@ public class EmbedBuilder
     {
         private readonly Embed _embed;
 
+        public const int MaxFieldCount = 25;
```
Any preference on using DiscordConfig for these consts?

moiph commented:
I preferred I can update all of the getters/setters to be consistent ( get => field  vs get { return field } ), usually I prefer to keep that type of cleanup isolated from functional changes.

AntiTcb commented on src/Discord.Net.Rest/Entities/Messages/EmbedBuilder.cs:
```diff
@@ -8,14 +8,36 @@ public class EmbedBuilder
+            }
+        }
+
         public string Url { get { return _embed.Url; } set { _embed.Url = value; } }
```
I can't help but feel like these would be better as `System.Uri`s rather than strings.

FiniteReality replied:
I concur; using the Uri type means that we get validation of the input as an artefact of its implementation, and IMO the Uri class is nicer to work with.

moiph replied:
Agreed I'm a fan of that change. So much for a single file edit :)

AntiTcb commented on src/Discord.Net.Rest/Entities/Messages/EmbedBuilder.cs:
```diff
@@ -149,8 +178,29 @@ public class EmbedFieldBuilder
+            get => _field.Name;
+            set
+            {
+                if (value?.Length > MaxFieldNameLength) throw new ArgumentException($"Field name length must be less than or equal to {MaxFieldNameLength}.");
```
Should probably also check field names/values for empty strings or nulls since those also throw back 400s.

moiph replied:
Good call, agreed

moiph committed:
Embed URI changes

Changes property types for any URLs in Embeds to System.URI.
Adding field name/value null/empty checks.

moiph commented:
Regarding the overall character limit -- it seems 4000 is not accurate.  So I'm intentionally excluding that for now until we can get a clearer idea of what those limits are.

FiniteReality reviewed (changes requested):
The ArgumentException instances should be passed the property name too, I missed this initially.

moiph committed:
including property names in argumentexceptions

moiph committed:
Adds overall embed length check

moiph commented on src/Discord.Net.Core/Entities/Messages/Embed.cs:
```diff
@@ -56,6 +57,8 @@ internal Embed(string type)
             Fields = fields;
         }
 
+        public int Length => Title?.Length + Author?.Name?.Length + Description?.Length + Footer?.Text?.Length + Fields.Sum(f => f.Name.Length + f.Value.ToString().Length) ?? 0;
```
This assumes only text properties count toward the limit; if URLs count then obviously this needs an update :)
